cedar-policy-core/src/est/expr.rs

```diff
@@ -21,34 +21,100 @@ use crate::entities::json::{
-    /// We try to match this first, see docs on #[serde(untagged)].
     ExprNoExt(ExprNoExt),
-    /// If that didn't match (because the key is not one of the keys defined in
-    /// `ExprNoExt`), we assume we have an extension function call, where the
-    /// key is the name of an extension function or method.
+    /// Extension function call, where the key is the name of an extension
+    /// function or method.
     ExtFuncCall(ExtFuncCall),
 }
 
+// Manual implementation of `Deserialize` is more efficient than the derived
+// implementation with `serde(untagged)`. In particular, if the key is valid for
+// `ExprNoExt` but there is a deserialization problem within the corresponding
+// value, the derived implementation would backtrack and try to deserialize as
+// `ExtFuncCall` with that key as the extension function name, but this manual
+// implementation instead eagerly errors out, taking advantage of the fact that
+// none of the keys for `ExprNoExt` are valid extension function names.
+//
+// See #1284.
+impl<'de> Deserialize<'de> for Expr {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct ExprVisitor;
+        impl<'de> Visitor<'de> for ExprVisitor {
+            type Value = Expr;
+            fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                formatter.write_str("JSON object representing an expression")
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::MapAccess<'de>,
+            {
+                let (k, v): (SmolStr, JsonValueWithNoDuplicateKeys) = match map.next_entry()? {
+                    None => {
+                        return Err(serde::de::Error::custom(
+                            "empty map is not a valid expression",
+                        ))
+                    }
+                    Some((k, v)) => (k, v),
+                };
+                match map.next_key()? {
+                    None => (),
+                    Some(k2) => {
+                        let k2: SmolStr = k2;
+                        return Err(serde::de::Error::custom(format!("JSON object representing an Expr should have only one key, but found two keys: `{k}` and `{k2}`")));
+                    }
+                };
+                match k.as_str() {
+                    // All the valid keys for `ExprNoExt`.
+                    // If a variant or alias is added to `ExprNoExt`, it must also be added here.
+                    "Value" | "Var" | "Slot" | "!" | "neg" | "==" | "!=" | "in" | "<" | "<="
+                    | ">" | ">=" | "&&" | "||" | "+" | "-" | "*" | "contains" | "containsAll"
+                    | "containsAny" | "getTag" | "hasTag" | "." | "has" | "like" | "is"
+                    | "if-then-else" | "Set" | "Record" => {
```

Comment:
I can see this coming back to bite us in the future. Is there a good reason not to have an `enum` with all the valid keys and use it here?

Reply:
1) I think the overall risk here is low.  If someone adds a case to `ExprNoExt` and not here, it will be caught by literally any EST deserialization test involving the new variant.  And no one will add a variant to `ExprNoExt` without also adding a test for it.
2) Putting these in an `enum` doesn't solve the problem of having to maintain the correspondence between the keys here and the keys of `ExprNoExt`.  But if you think having another `enum` would improve readability, what specifically is the proposal?  That we would try to deserialize `k` into that enum, instead of having the logic inline?

Reply:
We could also invert the logic -- have a list of known extension function names, and in those cases deserialize into the `ExtFuncCall` variant, in all other cases deserialize into the `ExprNoExt` variant.  Then we'd have to update on any addition of an extension function, rather than any addition to `ExprNoExt`.  My feeling was that additions of extension functions are likely to be more frequent.

Reply:
Actually, we might be able to use the `Extensions` object to dynamically look up whether a string is a valid extension function name, and route the deserialization accordingly.  This would rely on a single source of truth for defined extensions, and we wouldn't have to make updates in both places.  Would that be better?

Reply:
Yes, I'm fine with any alternative here.

It's true that adding an `enum` doesn't solve the problem, but adding just one `impl`  where we `match` on each variant of that `enum` should force updates to it when adding new ones. However, I haven't checked how `ExprNoExt` works in details so it's possible that wouldn't be so useful in the case of new keys.

Reply:
I like that solution. Though, we would ideally want to parametrize deserailzation on an `Extensions`, and I'm not sure how that would want.

Reply:
make sense. I guess then you should make sure we have a test case for parsing a deactivated but nonetheless existent extension function.

Reply:
Hmm actually, I think the current behavior is that a policy using a deactivated-but-nonetheless-existent extension function does parse (all the way to AST) and you only get an evaluation error

Reply:
Hmm this version with `is_known_extension_func_name()` is 20% slower on the 1284 benchmark than the original version (with the hardcoded match on list of ExprNoExt keys).  (now 129 µs on my machine.)  Not sure if we care since this is an edge case anyway.  More interested in if the happy-path performance changed

Reply:
ah, that's unfortunate. Maybe due to calling `ast::Name::parse_unqualified_name` a lot?
I think that still sounds like a (very) good improvement over current performance, so I'm still happy to see this merged.

Reply:
Yeah, it probably is due to that.  I wonder if there's a faster way to check this.

We could take advantage of the fact that currently no extension functions are namespaced, and have a function `is_unknown_extension_func_name()` that works on `Id` instead of `Name`.  That would avoid parsing.  We could even justify using `Id::new_unchecked`, because even if the name was reserved, it wouldn't match an extension func name